### PR TITLE
Refactor StateMachine to use Predicate<T> instead of Set<T> of terminal states

### PR DESCRIPTION
**Pull request by wendigo**
Required for https://github.com/prestosql/presto/pull/2584

### CODE REVIEW

**Comment by wendigo** [comment]
Done @findepi

File: presto-main/src/test/java/io/prestosql/execution/TestStateMachine.java [hunk 1]
```diff
@@ -132,11 +133,40 @@ public void testSet()
...
+    {
+        StateMachine<Optional<Throwable>> stateMachine = new StateMachine<>("test", executor, Optional.empty(), Optional::isPresent);
+        assertEquals(stateMachine.get(), Optional.empty());
+
+        Optional<Throwable> value = Optional.of(new Throwable("This is terminal state value"));
+        Optional<Throwable> value2 = Optional.of(new Throwable("This is other terminal state value"));
+
+        assertNoStateChange(stateMachine, () -> stateMachine.compareAndSet(value2, value));
+
+        assertStateChange(stateMachine,
```
Comment (findepi) [inline presto-main/src/test/java/io/prestosql/execution/TestStateMachine.java]: Per code-style, put all arguments on one line or each on separate line.
So this becomes:

```suggestion
        assertStateChange(
            stateMachine,
```

File: presto-main/src/test/java/io/prestosql/execution/TestStateMachine.java [hunk 2]
```diff
@@ -132,11 +133,40 @@ public void testSet()
...
+                value);
+
+        // attempt transition from a final state
+        assertNoStateChange(stateMachine, () -> {
+            try {
+                stateMachine.compareAndSet(value, value2);
+                fail("expected IllegalStateException");
+            }
+            catch (IllegalStateException expected) {
+            }
```
Comment (findepi) [inline presto-main/src/test/java/io/prestosql/execution/TestStateMachine.java]: you could validate the expected was raised:

`AtomicInteger expectedOccurred = new AtomicInteger();`

```suggestion
                expectedOccurred.incrementAndGet();
            }
```

+ `assertEquals(expectedOccurred.get(), 1);`

the existing tests can be improved too, as a separate preparatory commit

File: presto-main/src/main/java/io/prestosql/execution/StateMachine.java [hunk 3]
```diff
@@ -67,23 +66,23 @@
...
+     * @param isTerminalState predicate to test for terminal state
      */
-    public StateMachine(String name, Executor executor, T initialState, Iterable<T> terminalStates)
+    public StateMachine(String name, Executor executor, T initialState, Predicate<T> isTerminalState)
     {
         this.name = requireNonNull(name, "name is null");
         this.executor = requireNonNull(executor, "executor is null");
         this.state = requireNonNull(initialState, "initialState is null");
-        this.terminalStates = ImmutableSet.copyOf(requireNonNull(terminalStates, "terminalStates is null"));
+        this.terminalStatePredicate = requireNonNull(isTerminalState, "isTerminalState is null");
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/execution/StateMachine.java]: rename the field to `isTerminalState` too

File: presto-main/src/test/java/io/prestosql/execution/TestStateMachine.java (hunk 2 above)
Comment (wendigo) [reply presto-main/src/test/java/io/prestosql/execution/TestStateMachine.java]: Will do in n a separate commit / PR

**Pull request closed without merging**